pkg/arvo/sys/hoon.hoon

```diff
@@ -5868,180 +5868,205 @@
+            %d
           ?+    hay  (z-co q.p.lot)
-              $a
+              %a
             =+  yod=(yore q.p.lot)
-            =>  ^+(. .(rep ?~(f.t.yod rep ['.' (s-co f.t.yod)])))
-            =>  ^+  .
-                %=    .
-                    rep
-                  ?:  &(=(~ f.t.yod) =(0 h.t.yod) =(0 m.t.yod) =(0 s.t.yod))
-                    rep
-                  =>  .(rep ['.' (y-co s.t.yod)])
-                  =>  .(rep ['.' (y-co m.t.yod)])
-                  ['.' '.' (y-co h.t.yod)]
-                ==
-            =>  .(rep ['.' (a-co d.t.yod)])
-            =>  .(rep ['.' (a-co m.yod)])
-            =>  .(rep ?:(a.yod rep ['-' rep]))
+            =?  rep  ?=(^ f.t.yod)  ['.' (s-co f.t.yod)]
+            =?  rep  ?&  ?=(^ f.t.yod)
+                         !|(=(0 h.t.yod) =(0 m.t.yod) =(0 s.t.yod))
+                     ==
+              =.  rep  ['.' (y-co s.t.yod)]
+              =.  rep  ['.' (y-co m.t.yod)]
+              ['.' '.' (y-co h.t.yod)]
+            =.  rep  ['.' (a-co d.t.yod)]
+            =.  rep  ['.' (a-co m.yod)]
+            =?  rep  !a.yod  ['-' rep]
             ['~' (a-co y.yod)]
           ::
-              $r
+              %r
             =+  yug=(yell q.p.lot)
-            =>  ^+(. .(rep ?~(f.yug rep ['.' (s-co f.yug)])))
+            =?  rep  ?=(^ f.yug)  ['.' (s-co f.yug)]
             :-  '~'
             ?:  &(=(0 d.yug) =(0 m.yug) =(0 h.yug) =(0 s.yug))
               ['s' '0' rep]
-            =>  ^+(. ?:(=(0 s.yug) . .(rep ['.' 's' (a-co s.yug)])))
-            =>  ^+(. ?:(=(0 m.yug) . .(rep ['.' 'm' (a-co m.yug)])))
-            =>  ^+(. ?:(=(0 h.yug) . .(rep ['.' 'h' (a-co h.yug)])))
-            =>  ^+(. ?:(=(0 d.yug) . .(rep ['.' 'd' (a-co d.yug)])))
+            =?  rep  !=(0 s.yug)  ['.' 's' (a-co s.yug)]
+            =?  rep  !=(0 m.yug)  ['.' 'm' (a-co m.yug)]
+            =?  rep  !=(0 h.yug)  ['.' 'h' (a-co h.yug)]
+            =?  rep  !=(0 d.yug)  ['.' 'd' (a-co d.yug)]
             +.rep
           ==
         ::
-            $f
+            %f
           ?:  =(& q.p.lot)
             ['.' 'y' rep]
           ?:(=(| q.p.lot) ['.' 'n' rep] (z-co q.p.lot))
         ::
-            $n   ['~' rep]
-            $i
+            %n   ['~' rep]
+            %i
           ?+  hay  (z-co q.p.lot)
-            $f  ((ro-co [3 10 4] |=(a/@ ~(d ne a))) q.p.lot)
-            $s  ((ro-co [4 16 8] |=(a/@ ~(x ne a))) q.p.lot)
+            %f  ((ro-co [3 10 4] |=(a=@ ~(d ne a))) q.p.lot)
+            %s  ((ro-co [4 16 8] |=(a=@ ~(x ne a))) q.p.lot)
           ==
         ::
-            $p
+            %p
           =+  sxz=(fein:ob q.p.lot)
           =+  dyx=(met 3 sxz)
           :-  '~'
           ?:  (lte dyx 1)
             (weld (trip (tod:po sxz)) rep)
           =+  dyy=(met 4 sxz)
-          =+  imp=*@
+          =|  imp=@ud
           |-  ^-  tape
           ?:  =(imp dyy)
             rep
           %=  $
-            sxz  (rsh 4 1 sxz)
```

Comment:
`+end`/`+rsh` loops are pleasant to write and easy to understand, but decidedly suboptimal since they involve a bunch of incremental allocations. The practical choices are to either loop with a counter variable and use `+cut` (as this code now does), or `+rip`/`+ripn` the atom into a list and then map or reduce as needed. The former is definitely more efficient for large atoms, the latter is usually nicer code, and maybe more efficient for small-to-midsize atoms. We could use some more fundamental atom operations to make this kind of code nicer while being efficient.